File: harness/iterate-plugin/src/evidence.ts
/**
 * Deterministic code-evidence verification for review findings.
 *
 * Mirror of `iterate_harness/iterate/evidence.py` for the iterate-plugin.
 *
 * The iterate review loop requires that reviewer subagent findings ANCHOR to
 * real code instead of speculating. This module enforces it:
 *
 * - a finding's `file` must resolve to an existing file under the project root
 *   (traversal-safe), otherwise evidence is poisoned (`file_not_found`);
 * - a finding with an explicit line must reference a line that actually exists
 *   in that file (`line_out_of_range`);
 * - a whole-file finding (line 0 / undefined) must still reference an existing
 *   file, so even structural findings cannot point at nothing;
 * - `readVerified` is a best-effort, NON-gating hint: the plugin's reviewers are
 *   subagents whose reads are not aggregated here, so it is only set when a
 *   read set is explicitly provided and never fails the audit.
 *
 * Gate rule (user preference): ANY localizable finding with poisoned evidence
 * flips the whole audit to `passed: false`, so the meta-review forces revision.
 *
 * The pure math (`countLines`, `verifyLineBounds`) is separated from the
 * filesystem half (`verifyFinding`) to stay unit-testable without touching disk.
 */

import { existsSync, readFileSync, realpathSync, statSync } from 'node:fs'
import { resolve, sep } from 'node:path'
import type { ReviewFinding } from './types.ts'

/** Sentinel for whole-file findings (line 0 or omitted means the whole file). */
export const WHOLE_FILE_LINE = 0

/**
 * Hard cap on a single evidence file read. `verifyFinding` only needs the
 * line count + a NUL check; reading an unbounded file (or a device file
 * reached through a symlink) is a memory/hang hazard, so anything larger is
 * treated as not line-addressable.
 */
const MAX_EVIDENCE_BYTES = 10 * 1024 * 1024

export type EvidenceError = 'file_not_found' | 'line_out_of_range'

/** Per-finding attestation result. */
export interface FindingEvidence {
  file: string
  line: number | null
  lineTotal: number | null
  resolvedPath: string | null
  verified: boolean
  error?: EvidenceError
  /** True/False only when a read-set is supplied; undefined = not checkable. */
  readVerified?: boolean
}

/** Aggregate attestation over a findings list. */
export interface EvidenceAudit {
  checked: number
  results: FindingEvidence[]
}

/** A single finding object that exposes `file` / `line` (for verification). */
interface Locatable {
  file?: string
  line?: number
}

/** Number of physical lines in `text`. A trailing newline does not add a line. */
export function countLines(text: string): number {
  if (text === '') return 0
  // Mirrors Python `str.splitlines()`: split on every line separator, not just
  // \r\n|\r|\n — otherwise line counts diverge from the harness on files
  // containing \v \f \x1c-\x1e \x85 \u2028 \u2029.
  const parts = text.split(/\r\n|[\n\r\v\f\x1c\x1d\x1e\x85\u2028\u2029]/)
  // A trailing newline leaves an empty final element that is NOT a line
  // (mirrors Python `str.splitlines()` used by the harness).
  if (parts[parts.length - 1] === '') return parts.length - 1
  return parts.length
}

/** Resolve `root/rel` and reject any path escaping `root` (returns null). */
export function resolveWithin(root: string, rel: string): string | null {
  const resolved = resolve(root, rel)
  const rootResolved = resolve(root)
  if (resolved === rootResolved) return resolved
  const prefix = rootResolved.endsWith(sep) ? rootResolved : rootResolved + sep
  if (!resolved.startsWith(prefix)) return null
  return resolved
}

/** True when `candidate` is `root` itself or lexically inside `root`. */
function isWithin(root: string, candidate: string): boolean {
  if (candidate === root) return true
  const prefix = root.endsWith(sep) ? root : root + sep
  return candidate.startsWith(prefix)
}

/** best-effort realpath; falls back to the lexical path on any failure. */
function safeRealpath(p: string): string {
  try {
    return realpathSync(p)
  } catch {
    return p
  }
}

/**
 * Pure check that `line` (if anchored) exists in `text`.
 * Whole-file findings (undefined/0) are always bounds-valid.
 */
export function verifyLineBounds(
  line: number | null | undefined,
  text: string,
): { inBounds: boolean; lineTotal: number } {
  const lineTotal = countLines(text)
  if (line === undefined || line === null || line === WHOLE_FILE_LINE) {
    return { inBounds: true, lineTotal }
  }
  if (line < 1) return { inBounds: false, lineTotal }
  return { inBounds: line <= lineTotal, lineTotal }
}

/** Verify a single finding's location against the real filesystem. */
export function verifyFinding(
  root: string,
  input: Locatable,
  opts: { readSet?: Set<string> } = {},
): FindingEvidence {
  const relFile = input.file ?? ''
  const line = typeof input.line === 'number' ? input.line : null
  const resolved = resolveWithin(root, relFile)

  if (resolved === null || !existsSync(resolved)) {
    return {
      file: relFile,
      line,
      lineTotal: null,
      resolvedPath: resolved,
      verified: false,
      error: 'file_not_found',
    }
  }

  // Symlink containment: resolveWithin is lexical only, but existsSync /
  // readFileSync follow symlinks. Verify the REAL path stays inside the REAL
  // project root so a finding path can never read (or line-count) a file
  // outside the project via a symlinked directory or file.
  const rootReal = safeRealpath(root)
  const real = safeRealpath(resolved)
  if (!isWithin(rootReal, real)) {
    return {
      file: relFile,
      line,
      lineTotal: null,
      resolvedPath: resolved,
      verified: false,
      error: 'file_not_found',
    }
  }

  // Regular-file + size guard: a directory, device file (/dev/zero), FIFO or
  // multi-GB file is not a line-addressable text target. statSync follows
  // symlinks, so a link to a device still lands here and is rejected.
  let st
  try {
    st = statSync(resolved)
  } catch {
    return {
      file: relFile,
      line,
      lineTotal: null,
      resolvedPath: resolved,
      verified: false,
      error: 'file_not_found',
    }
  }
  if (!st.isFile() || st.size > MAX_EVIDENCE_BYTES) {
    return {
      file: relFile,
      line,
      lineTotal: null,
      resolvedPath: resolved,
      verified: false,
      error: 'line_out_of_range',
    }
  }

  let raw: Buffer
  try {
    raw = readFileSync(resolved)
  } catch {
    return {
      file: relFile,
      line,
      lineTotal: null,
      resolvedPath: resolved,
      verified: false,
      error: 'file_not_found',
    }
  }

  // A file is not line-addressable if it contains a NUL byte (binary payload).
  // Anchored line numbers on a binary file cannot be trusted, so treat them the
  // same as an out-of-range line rather than credulously accepting them
  // (mirrors the harness `evidence.py` NUL check).
  if (raw.includes(0)) {
    return {
      file: relFile,
      line,
      lineTotal: null,
      resolvedPath: resolved,
      verified: false,
      error: 'line_out_of_range',
    }
  }

  const text = raw.toString('utf-8')
  const { inBounds, lineTotal } = verifyLineBounds(line, text)
  if (!inBounds) {
    return {
      file: relFile,
      line,
      lineTotal,
      resolvedPath: resolved,
      verified: false,
      error: 'line_out_of_range',
    }
  }

  const outcome: FindingEvidence = {
    file: relFile,
    line,
    lineTotal,
    resolvedPath: resolved,
    verified: true,
  }
  if (opts.readSet !== undefined) {
    outcome.readVerified = opts.readSet.has(resolved)
  }
  return outcome
}

/** Attest every finding in a list. */
export function verifyFindings(
  root: string,
  findings: Locatable[],
  opts: { readSet?: Set<string> } = {},
): EvidenceAudit {
  const results = findings.map((f) => verifyFinding(root, f, opts))
  return { checked: results.length, results }
}

/** `passed` is true only when no real existence failure exists (read is a hint). */
export function evidencePassed(audit: EvidenceAudit): boolean {
  return audit.results.every((r) => r.error === undefined)
}

/** Violating (non-grounded) results. */
export function evidenceViolations(audit: EvidenceAudit): FindingEvidence[] {
  return audit.results.filter((r) => r.error !== undefined)
}

/** Serialize an audit for tool payloads (pure). */
export function evidenceToPlain(audit: EvidenceAudit): Record<string, unknown> {
  const computable = audit.results.filter((r) => r.readVerified !== undefined)
  const readRatio =
    computable.length === 0
      ? null
      : Number(
          (
            computable.filter((r) => r.readVerified === true).length / computable.length
          ).toFixed(3),
        )
  return {
    checked: audit.checked,
    passed: evidencePassed(audit),
    violations: audit.results
      .filter((r) => r.error !== undefined)
      .map((r) => ({ file: r.file, line: r.line, lineTotal: r.lineTotal, verified: r.verified, error: r.error })),
    readVerifiedRatio: readRatio,
  }
}

File: harness/iterate-harness/src/iterate_harness/iterate/evidence.py
"""Deterministic code-evidence verification for review findings.

The iterate review loop requires that reviewer subagents ANCHOR every finding
to real code instead of speculating. This module makes that requirement
enforceable, not just a prompt exhortation:

- a finding's ``file`` must resolve to an existing file under the scope root,
  otherwise the evidence is poisoned (``file_not_found``) — this kills
  fabricated file paths;
- a finding with an explicit line must reference a line that actually exists
  within that file (``line_out_of_range``) — this kills invented line numbers;
- a whole-file finding (``line`` is ``None`` or ``0``) must still reference an
  existing file, so even structural findings cannot point at nothing;
- ``read_verified`` is a best-effort, NON-gating hint that the file appeared
  in the current session's ``read_file`` carry-over. It is never used to fail
  the audit because the per-dimension reviewer reads happen inside subagents
  whose tool calls are not aggregated onto the main context at this point.

The gate rule is deliberately strict (user preference): ANY localizable
finding whose evidence fails flips the whole attestation to ``passed=False``.

The file-system half (path resolution + line counting) is separated from the
pure math so the module can be unit-tested without touching disk.

This module is an I/O-capable companion to the pure :mod:`.meta_review`; it
contains no agent spawning.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

#: Sentinel for whole-file findings (`line` unset or 0 means the whole file).
WHOLE_FILE_LINE = 0

EvidenceError = Literal["file_not_found", "line_out_of_range"]


def count_lines(text: str) -> int:
    """Number of physical lines in ``text``.

    Uses :meth:`str.splitlines` so ``""`` is 0 and a trailing newline does not
    fabricate an extra line (``"a\\nb\\n"`` is 2, not 3).
    """
    if text == "":
        return 0
    return len(text.splitlines())


@dataclass(frozen=True)
class FindingEvidence:
    """Per-finding attestation result."""

    file: str
    line: int | None
    line_total: int | None
    resolved_path: str | None
    verified: bool
    error: EvidenceError | None = None
    #: True/False only when a read-set is supplied; None = not checkable.
    read_verified: bool | None = None

    def to_dict(self) -> dict[str, object]:
        out: dict[str, object] = {
            "file": self.file,
            "line": self.line,
            "lineTotal": self.line_total,
            "resolvedPath": self.resolved_path,
            "verified": self.verified,
        }
        if self.error is not None:
            out["error"] = self.error
        if self.read_verified is not None:
            out["readVerified"] = self.read_verified
        return out


@dataclass
class EvidenceAudit:
    """Aggregate attestation over a findings list."""

    checked: int
    results: list[FindingEvidence] = field(default_factory=list)

    @property
    def passed(self) -> bool:
        # read_verified is a hint only; only real existence failures gate.
        return all(r.error is None for r in self.results)

    @property
    def errors(self) -> list[FindingEvidence]:
        return [r for r in self.results if r.error is not None]

    def to_dict(self) -> dict[str, object]:
        return {
            "checked": self.checked,
            "passed": self.passed,
            "violations": [r.to_dict() for r in self.errors],
            "readVerifiedRatio": _ratio(self),
        }


def _ratio(audit: EvidenceAudit) -> float | None:
    computable = [r for r in audit.results if r.read_verified is not None]
    if not computable:
        return None
    return round(sum(1 for r in computable if r.read_verified) / len(computable), 3)


def resolve_within(root: Path, rel: str) -> Path | None:
    """Resolve ``root / rel`` and reject any path escaping ``root``.

    Returning ``None`` on traversal or on unresolvable paths keeps every
    subsequent existence check safe against path-escape.
    """
    try:
        resolved = (root / rel).resolve()
    except OSError:
        return None
    try:
        root_resolved = root.resolve()
    except OSError:
        return None
    if resolved == root_resolved:
        return resolved
    if root_resolved not in resolved.parents:
        return None
    return resolved


def verify_line_bounds(line: int | None, text: str) -> tuple[bool, int]:
    """Pure check that ``line`` (if anchored) exists in ``text``.

    Whole-file findings (``line is None`` or ``0``) are always bounds-valid — a
    structural finding only needs its file to exist. Returns ``(in_bounds,
    line_total)``.
    """
    total = count_lines(text)
    if line is None or line == WHOLE_FILE_LINE:
        return True, total
    if line < 1:
        return False, total
    return line <= total, total


def verify_finding(
    root: Path,
    *,
    rel_file: str,
    line: int | None,
    read_set: set[str] | None = None,
) -> FindingEvidence:
    """Verify a single finding's location against the real filesystem.

    ``read_set`` carries the current session's unique resolved read paths; it is
    only used to fill the non-gating ``read_verified`` hint.
    """
    resolved = resolve_within(root, rel_file)
    if resolved is None or not resolved.is_file():
        return FindingEvidence(
            file=rel_file,
            line=line,
            line_total=None,
            resolved_path=str(resolved) if resolved is not None else None,
            verified=False,
            error="file_not_found",
        )

    try:
        raw = resolved.read_bytes()
    except OSError:
        # Unreadable (permissions, broken link...) — treat as missing evidence.
        return FindingEvidence(
            file=rel_file,
            line=line,
            line_total=None,
            resolved_path=str(resolved),
            verified=False,
            error="file_not_found",
        )

    # A file is not line-addressable if it contains a NUL byte (binary payload).
    # Anchored line numbers on a binary file cannot be trusted, so treat them
    # the same as an out-of-range line rather than credulously accepting them.
    if b"\x00" in raw:
        return FindingEvidence(
            file=rel_file,
            line=line,
            line_total=None,
            resolved_path=str(resolved),
            verified=False,
            error="line_out_of_range",
        )

    text = raw.decode("utf-8", errors="replace")
    in_bounds, total = verify_line_bounds(line, text)
    if not in_bounds:
        return FindingEvidence(
            file=rel_file,
            line=line,
            line_total=total,
            resolved_path=str(resolved),
            verified=False,
            error="line_out_of_range",
        )

    read_verified: bool | None = None
    if read_set is not None:
        read_verified = _normalized(resolved) in read_set

    return FindingEvidence(
        file=rel_file,
        line=line,
        line_total=total,
        resolved_path=str(resolved),
        verified=True,
        read_verified=read_verified,
    )


def _normalized(path: Path) -> str:
    try:
        return os.path.normcase(str(path.resolve()))
    except OSError:
        return os.path.normcase(str(path))


def verify_findings(
    root: Path,
    *,
    findings: list[object],
    read_set: set[str] | None = None,
) -> EvidenceAudit:
    """Attest every finding in a list (any object exposing ``file``/``line``)."""
    results: list[FindingEvidence] = []
    for finding in findings:
        rel_file = str(getattr(finding, "file", ""))
        line = getattr(finding, "line", None)
        results.append(
            verify_finding(
                root, rel_file=rel_file, line=line, read_set=read_set
            )
        )
    return EvidenceAudit(checked=len(results), results=results)


def read_set_from_metadata(metadata: dict[str, object] | None) -> set[str] | None:
    """Normalize the session's ``read_file_state`` carry-over into a path set.

    Returns ``None`` when no read trace is present (so callers skip the
    non-gating hint rather than falsely reporting every file as unread).
    """
    if not metadata:
        return None
    raw = metadata.get("read_file_state")
    if not isinstance(raw, list) or not raw:
        return None
    paths: set[str] = set()
    for entry in raw:
        if not isinstance(entry, dict):
            continue
        path = entry.get("path")
        if isinstance(path, str) and path:
            paths.add(os.path.normcase(path))
    return paths or None